### PR TITLE
Emulated and ed25519 crates

### PR DESCRIPTION
**Pull request by avras**
Adding two crates

- `bellpepper-emulated` is a crate for implementing non-native arithmetic. It is inspired by the inspired by the [emulated](https://github.com/Consensys/gnark/tree/master/std/math/emulated) package in [Gnark](https://github.com/Consensys/gnark). It has been tested via an implementation of `ed25519` curve operations. Further testing through implementation of other curves like `secp256k1` is a TODO.
- `bellpepper-ed25519` is a crate for implementing ed25519 curve operations. It does NOT support ed25519 signature verification as of now. Only point addition and scalar multiplication are implemented.

### CODE REVIEW

File: crates/bellpepper-ed25519/src/circuit.rs [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+        assert!(sum_alloc.is_ok());
+        let sum_al = sum_alloc.unwrap();
+
+        assert_eq!(sum_expected_value, sum_al.value);
+
+        if !cs.is_satisfied() {
+            eprintln!("{:?}", cs.which_is_unsatisfied())
+        }
+        assert!(cs.is_satisfied());
+        println!("Num constraints = {:?}", cs.num_constraints());
```
Comment (emmorais) [inline crates/bellpepper-ed25519/src/circuit.rs:517]: Suggestion: use `assert_eq` to validate the number of constraints. This is happening in other places of the code base and turned out to be useful way to inform the code reader.

File: crates/bellpepper-ed25519/src/circuit.rs [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+        println!("Num inputs = {:?}", cs.num_inputs());
```
Comment (emmorais) [inline crates/bellpepper-ed25519/src/circuit.rs:518]: Same for number of inputs.

File: crates/bellpepper-ed25519/src/circuit.rs [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+        assert!(double_alloc.is_ok());
+        let double_al = double_alloc.unwrap();
+
+        assert_eq!(double_expected_value, double_al.value);
+
+        if !cs.is_satisfied() {
+            eprintln!("{:?}", cs.which_is_unsatisfied())
+        }
+        assert!(cs.is_satisfied());
+        println!("Num constraints = {:?}", cs.num_constraints());
```
Comment (emmorais) [inline crates/bellpepper-ed25519/src/circuit.rs:547]: Replace the `println` by `assert_eq`.

File: crates/bellpepper-ed25519/src/circuit.rs [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+            &mut cs.namespace(|| "scalar multiplication"),
+            scalar_vec,
+        );
+        assert!(p_alloc.is_ok());
+        let p_al = p_alloc.unwrap();
+
+        assert_eq!(p, p_al.value);
+
+        assert!(cs.is_satisfied());
+        println!("Num constraints = {:?}", cs.num_constraints());
```
Comment (emmorais) [inline crates/bellpepper-ed25519/src/circuit.rs:589]: Replace the `println` by `assert_eq`.

File: crates/bellpepper-ed25519/src/circuit.rs [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+            scalar_vec,
+            window_size,
+        );
+        assert!(p_alloc.is_ok());
+        let p_al = p_alloc.unwrap();
+
+        assert_eq!(p, p_al.value);
+
+        assert!(cs.is_satisfied());
+        println!(
```
Comment (emmorais) [inline crates/bellpepper-ed25519/src/circuit.rs:640]: Replace the `println` by `assert_eq`.

**Comment by avras** [comment]
I am fixing the `cargo clippy` warnings. Will commit the changes soon.

File: crates/bellpepper-ed25519/src/circuit.rs (hunk 1 above)
Comment (avras) [reply crates/bellpepper-ed25519/src/circuit.rs:517]: Will do in all such locations. Thanks.

Comment (avras) [reply crates/bellpepper-ed25519/src/circuit.rs]: @emmorais I have replaced the println's with assert_eq in all locations. Please let me know if you have other comments/suggestions.

**Review by emmorais — APPROVED** [review]

**Review by huitseeker — APPROVED** [review]
Thank you so much for the excellent contribution @avras and @varunthakore!

**Pull request merged**